### PR TITLE
Drupal 9

### PR DESCRIPTION
**Pull request by piggito**
Update drupal core to v9 including all required dependencies.

We need to consider the issues previously reported in https://github.com/skilld-labs/skilld-docker-container/pull/233 here as well

### CODE REVIEW

File: .env.default [hunk 1]
```diff
@@ -1,7 +1,7 @@
 COMPOSE_FILE=./docker/docker-compose.yml:./docker/docker-compose.override.yml
 COMPOSE_PROJECT_NAME=projectname
 PROFILE_NAME=sdd
-MODULES=project_default_content better_normalizers default_content hal serialization
+MODULES=project_default_content default_content serialization
```
Comment (piggito) [inline .env.default:4]: Why were better_normalizers and hal removed?

File: .env.default [hunk 2]
```diff
@@ -21,3 +21,5 @@ DB_URL=sqlite:./../.cache/d8.sqlite
 #DB_URL=sqlite:///dev/shm/d8.sqlite # Faster but data will be lost on php container recreation
 #DB_URL=mysql://d8:d8@mysql/d8
 DB_DATA_DIR=../.cache # Include path to this folder to your .gitignore if you override it
+
+INSTALL_DEV_DEPENDENCIES=TRUE
```
Comment (piggito) [inline .env.default:25]: this seems unnecessary

File: composer.json [hunk 3]
```diff
@@ -28,47 +28,28 @@
...
   },
   "require-dev": {
     "dmore/behat-chrome-extension": "^1.3",
     "drupal/config_inspector": "^1",
+    "drupal/default_content": "1.x-dev",
     "drupal/devel": "3.x-dev",
-    "drupal/drupal-extension": "^3.4",
+    "drupal/manage_display": "1.x-dev",
+    "drupal/page_manager": "4.x-dev",
+    "drupal/redirect": "1.x-dev",
```
Comment (piggito) [inline composer.json:52]: None of these modules should be moved to dev requirements

File: composer.json [hunk 4]
```diff
@@ -88,25 +69,43 @@
...
   "scripts": {
     "drupal-scaffold": "DrupalComposer\\DrupalScaffold\\Plugin::scaffold",
-    "create-required-files": "SkilldDrupal\\composer\\ScriptHandler::createRequiredFiles",
     "list-scaffold-files": [
       "SkilldDrupal\\composer\\ScriptHandler::listScaffoldFiles"
     ],
     "pre-install-cmd": [
       "SkilldDrupal\\composer\\ScriptHandler::checkComposerVersion"
     ],
+    "create-required-files": "SkilldDrupal\\composer\\ScriptHandler::createRequiredFiles",
```
Comment (piggito) [inline composer.json:78]: No need to move these lines

File: composer.json [hunk 5]
```diff
@@ -88,25 +69,43 @@
...
+      ],
+      "web/modules/contrib/{$name}": [
+        "type:drupal-module"
+      ],
+      "web/profiles/contrib/{$name}": [
+        "type:drupal-profile"
+      ],
+      "web/themes/contrib/{$name}": [
+        "type:drupal-theme"
+      ],
```
Comment (piggito) [inline composer.json:99]: No need to change formatting here

File: composer.json [hunk 6]
```diff
@@ -88,25 +69,43 @@
...
+      "drush/Commands/contrib/{$name}": [
+        "type:drupal-drush"
+      ],
```
Comment (piggito) [inline composer.json:102]: Do we really need this? I don't remember using any `drupal-drush` package

File: composer.json [hunk 7]
```diff
@@ -88,25 +69,43 @@
...
+      "web/modules/custom/{$name}": [
+        "type:drupal-custom-module"
+      ],
+      "web/themes/custom/{$name}": [
+        "type:drupal-custom-theme"
+      ]
```
Comment (piggito) [inline composer.json:108]: AFAIK every drupal module or theme installed with composer is contrib so there won't be need for these lines

File: composer.json [hunk 8]
```diff
@@ -116,43 +115,17 @@
         "[web-root]/web.config": false
       }
     },
+    "drupal-core-vendor-hardening": {
+      "drupal/module_name": ["tests", "src/Tests"]
+    },
+    "create-required-files": "SkilldDrupal\\composer\\ScriptHandler::createRequiredFiles",
```
Comment (piggito) [inline composer.json:121]: This is duplicate from https://github.com/skilld-labs/skilld-docker-container/pull/243/files#diff-b5d0ee8c97c7abd7e3fa29b9a27d1780R78

File: web/modules/custom/project_default_content/project_default_content.info.yml [hunk 9]
```diff
@@ -4,9 +4,9 @@ package: Custom
 
 type: module
 core: 8.x
+core_version_requirement: ^8||^9
 
 dependencies:
   - default_content:default_content
-  - better_normalizers:better_normalizers
```
Comment (piggito) [inline web/modules/custom/project_default_content/project_default_content.info.yml:10]: Why was this dependency removed?

File: scripts/composer/ScriptHandler.php [hunk 10]
```diff
@@ -12,6 +12,7 @@
 use DrupalFinder\DrupalFinder;
 use Symfony\Component\Filesystem\Filesystem;
 use Webmozart\PathUtil\Path;
+use Drupal\Core\Site\Settings;
```
Comment (piggito) [inline scripts/composer/ScriptHandler.php:15]: Keep imports ordered

File: composer.json [hunk 11]
```diff
@@ -28,18 +28,18 @@
     }
   ],
   "require": {
-    "php": ">=7.0.8",
     "ckeditor/panelbutton": "^4.10",
+    "composer/installers": "^1.2",
```
Comment (andypost) [inline composer.json:32]: This one is dependency od recommended, also it is under fire in https://www.drupal.org/project/drupal/issues/3134648

File: composer.json [hunk 12]
```diff
@@ -28,18 +28,18 @@
...
-    "drupal-composer/drupal-scaffold": "^2.5",
+    "drupal/core-composer-scaffold": "^9",
     "drupal/better_normalizers": "^1.0@beta",
     "drupal/block_content_permissions": "^1.8",
-    "drupal/components": "^1",
-    "drupal/config_installer": "^1",
-    "drupal/core-recommended": "^8.8.0",
+    "drupal/components": "^2",
+    "drupal/core-recommended": "^9.0",
+    "drupal/core-vendor-hardening": "^8.8",
```
Comment (andypost) [inline composer.json:39]: Why not not use `^9.0` for all core provided components?

File: composer.json [hunk 13]
```diff
@@ -116,6 +113,9 @@
         "[web-root]/web.config": false
       }
     },
+    "drupal-core-vendor-hardening": {
+      "drupal/module_name": ["tests", "src/Tests"]
```
Comment (andypost) [inline composer.json:117]: The module name should be real one

Comment (andypost) [reply composer.json:117]: At least core tests & umami profile could be removed

Comment (piggito) [reply composer.json:117]: For this to work we need to create a patch for 9.0 in https://www.drupal.org/project/drupal/issues/3108262

File: composer.json (hunk 11 above)
Comment (piggito) [reply composer.json]: Removed as it is dependency of another package

File: composer.json (hunk 12 above)
Comment (piggito) [reply composer.json]: Changed all core packages to `^9.0` as suggested

File: composer.json (hunk 13 above)
Comment (piggito) [reply composer.json]: Actually we can't apply patches in d/o issue cause composer uses a mirror github repo at https://github.com/drupal/core-vendor-hardening which only includes a subfolder of Drupal's gitlab repo so I will be removing this extra config.
We can add removal for core tests and umami once d/o issue gets fixed.

**Comment by piggito** [comment]
Rebased and updated for install without config_installer.

File: scripts/makefile/tests.mk [hunk 14]
```diff
@@ -56,7 +56,8 @@ endif
 ## Validate composer.json file
 compval:
 	@echo "Composer.json validation..."
-	@docker run --rm -v `pwd`:`pwd` -w `pwd` $(IMAGE_PHP) composer validate --strict
+	# Can't use --strict cause we need dev versions for d9 compatibility
```
Comment (davidferlay) [inline scripts/makefile/tests.mk:59]: Need to remove it from there too https://github.com/skilld-labs/skilld-docker-container/blob/d9/.gitlab-ci.yml#L49

Comment (piggito) [reply scripts/makefile/tests.mk:59]: Fixed

**Comment by andypost** [comment]
Needs to explore, probably we should manage composer/installers top level https://git.drupalcode.org/project/drupal/-/commit/fbdf57f4ae8a932798fd453158ec128ee6d08c31

**Comment by davidferlay** [comment]
LGTM !

**Pull request merged**